### PR TITLE
[FIX] website: translate configurator's features page

### PR DESCRIPTION
**Pull request by detrouxdev**
Prior to this commit the configurator's features and pages were
not translated.

This commit fixes that by using the right env to make call to the
website.configurator.features model

task-2687416

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website/models/website.py [hunk 1]
```diff
@@ -292,7 +292,7 @@ def _website_api_rpc(self, route, params):
     def configurator_init(self):
         r = dict()
         company_id = self.get_current_website().company_id
-        configurator_features = self.env['website.configurator.feature'].search([])
+        configurator_features = self.env['website.configurator.feature'].with_context(lang=self.env.user.lang).search([])
```
Comment (qsm-odoo) [inline addons/website/models/website.py:295]: Seems louch, is not this supposed to be automatic?

File: addons/website/models/website.py [hunk 2]
```diff
@@ -292,7 +292,7 @@ def _website_api_rpc(self, route, params):
     def configurator_init(self):
         r = dict()
         company_id = self.get_current_website().company_id
-        configurator_features = self.env['website.configurator.feature'].search([])
+        configurator_features = request.env['website.configurator.feature'].search([])
```
Comment (qsm-odoo) [inline addons/website/models/website.py:295]: @detrouxdev Maybe you already decided this with @rdeodoo but should not you also fix the other `self.env` uses in the configurator code? I see one in the same function here with the call to `_website_api_rpc`, and that method also uses `self.env`

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/models/website.py (hunk 2 above)
Comment (detrouxdev) [reply addons/website/models/website.py:295]: I guess I see no reason not to change `self.env.user.lang` to `request.env.user.lang` but I was cautious to change it in other places as I'm not sure I understand when to use self.env instead request.env. Is there any moments where request would not be available?

Comment (rdeodoo) [reply addons/website/models/website.py:295]: If you are inside a controller: use request
If you are inside a model method which is called through RPC only (like this, or simply a code encapsulation inside controllers/main.py which is only called by multiple controllers eg `_get_products_recently_viewed()` ): use request
Otherwise: use self, request might be unbound (it comes from http.py, see `request = _request_stack()`)

**Comment by detrouxdev** [comment]
@rdeodoo should I replace every `self.env` by `request.env` in the `configurator_apply` method? It seemed like a big diff to me but I tried it and there was not apparent bugs.

I also did 
```python
if request:
    self.env = request.env
```
which worked but seemed not ideal.

File: addons/website/models/website.py [hunk 3]
```diff
@@ -479,6 +479,10 @@ def set_images(images):
         theme_name = kwargs['theme_name']
         theme = self.env['ir.module.module'].search([('name', '=', theme_name)])
         url = theme.button_choose_theme()
+        # Redirect to the default language of the website
+        # otherwise the user could get stuck in a loop
```
Comment (rdeodoo) [inline addons/website/models/website.py:483]: It is not really a loop, the comment is misleading. It is just the loading animation never being removed as the code is never catching the `enable_editor=1` correctly as there is no Edit button in translation mode, thus the Edit Mode is never started

**Comment by rdeodoo** [comment]
> I also did [...] which worked but seemed not ideal.

Indeed, not ideal.

> should I replace every self.env by request.env in the configurator_apply method?

Yes, as this is a method only called through RPC, you should replace the self.env by request.env

File: addons/website/models/website.py [hunk 4]
```diff
@@ -477,23 +477,31 @@ def set_images(images):
...
+        theme = request.env['ir.module.module'].search([('name', '=', theme_name)])
         url = theme.button_choose_theme()
+        # button_choose_theme redirects the user with
+        # with_loading and enable_editor but if the
+        # user is on a translated page (e.g /en)
+        # the editor never starts (as there's no edit button)
+        # thus never removes the loading animation
+        # so we redirect the user to the default 
+        # language so the editor can initialize properly
+        url['url'] = '/website/lang/default?r=%s' % urls.url_quote_plus(url['url'])
```
Comment (xmo-odoo) [inline addons/website/models/website.py:489]: :+1: 

Should use `url_encode` though, that's what we normally use unless there are specific reasons not to.

File: addons/website/models/ir_http.py [hunk 5]
```diff
@@ -443,6 +443,9 @@ def get_frontend_session_info(self):
             session_info.update({
                 'website_id': request.website.id,
                 'website_company_id': request.website._get_cached('company_id'),
+                # We need the user_context if we want the configurator
```
Comment (rdeodoo) [inline addons/website/models/ir_http.py:446]: Trailing whitespace here, please check your linter :D 

image

(last line)
image

File: addons/website/models/ir_http.py [hunk 6]
```diff
@@ -443,6 +443,9 @@ def get_frontend_session_info(self):
             session_info.update({
                 'website_id': request.website.id,
                 'website_company_id': request.website._get_cached('company_id'),
+                # We need the user_context if we want the configurator 
+                # to be in the user's language.
+                'user_context': request.session.get_context(),
```
Comment (rdeodoo) [inline addons/website/models/ir_http.py:448]: About this commit, the code of the configurator seems to be so unusual, a bit of technical explanation might be useless here.. Or I/we have to redo all the technical investigation you probably did to figure stuff and ensure the fix is right.

In this case, correct me if I am wrong:
- We have a website=True controller, which then need to add a lang to its call to `render()` to avoid using the frontend lang (frontend lang is obviously expected as it is a website=True controller, so a frontend lang) 
- That controller returns an empty page (empty body) which is then set in JS (not only the dynamic content, but the static sentences too like "Ready to build your website?" :thinking: )

So you need to do this kind of black magic (which is a generic fix used only for this weird controller, right?) to alter the Odoo normal behavior of the frontend, to translate the content it gets through RPC in another language than the one of the frontend.
It just seems like we are on a frontend controller but actually just want it to act as a backend controller :shrug: 

I wonder why this `/website/configurator` controller is set as frontend and actually inherit the frontend_layout instead of just being a regular backend controller.
```xml
<template id="website_configurator" name="Website Configurator" inherit_id="web.frontend_layout" primary="True">
    <xpath expr="//head/t[@t-call-assets][last()]" position="after">
        <t t-call-assets="website.website_configurator_assets_js" lazy_load="True"/>
    </xpath>
    <xpath expr="//t[@t-call-assets='web.assets_frontend']" position="replace">
        <t t-call-assets="website.website_configurator_assets_scss" t-js="false"/>
    </xpath>
    <xpath expr="//nav" position="replace"/>
    <xpath expr="//div[@id='wrapwrap']" position="replace"/>
</template>
```
There should be an obvious reason, but I really never dig into the configurator code.

Not saying your fix is wrong, just trying to figure stuff here.

Anyway, to sum up your fix, you need it for `localizationService`, which depends of `userService`, to get the wanted lang, from the context?

```js
get lang() {
    return context.lang;
},
```

**Review by rdeodoo — COMMENTED** [review]
The configurator is very dependent of the dispatcher/lang behavior.
With the new http-pocalypse coming, I am afraid such behaviors might be lost, as it really impacts those layer.
Couldn't you write some tests here?
I am not sure if you can robustly test those fix with python unit test (calling controller or method with MockRequest I guess?) or if you need to make a tour.
But that should definitely be part of the PR, IMHO.

File: addons/website/models/ir_http.py (hunk 6 above)
Comment (detrouxdev) [reply addons/website/models/ir_http.py:448]: The summarisation is correct. I needed the lang in the context for the localizationService to work properly. 

I'm not sure I understand what Website=true does. I tried looking in the code but I couldn't find a direct answer. Do you happen to know where I could find infos about that? I could then check the configurator code and have an answer as to why it's frontend.

In any case, it might be interesting to move it to the backend in master if task-2687506 gets merged (move edit to backend), because with the current code, I think the user would : create a new website / install from backend -> move to front for the configurator -> move to backend to edit.

**Comment by rdeodoo** [comment]
> > should I replace every self.env by request.env in the configurator_apply method?
> 
> Yes, as this is a method only called through RPC, you should replace the self.env by request.env

Do you remember the one line that really **needed** to be fixed? Could you add a comment on it on the PR? Thanks :+1:

**Comment by rdeodoo** [comment]
Could you add a clear step to reproduce (list type) on each commit please?

File: addons/website/models/website.py [hunk 7]
```diff
@@ -479,6 +479,14 @@ def set_images(images):
...
         theme = request.env['ir.module.module'].search([('name', '=', theme_name)])
         url = theme.button_choose_theme()
+        # button_choose_theme redirects the user with
+        # with_loading and enable_editor but if the
+        # user is on a translated page (e.g /en)
+        # the editor never starts (as there's no edit button)
+        # thus never removes the loading animation
+        # so we redirect the user to the default 
+        # language so the editor can initialize properly
+        url['url'] = '/website/lang/default?%s' % urls.url_encode({'r': url['url']})
```
Comment (JKE-be) [inline addons/website/models/website.py:489]: should be fixed at the root cause ;) and not only where it causes problem for your use case...
-> the url generation...
```diff
 addons/website/models/website.py | 10 ++--------
 1 file changed, 2 insertions(+), 8 deletions(-)

diff --git addons/website/models/website.py addons/website/models/website.py
index 63bf05fea348..ac921af4c41e 100644
--- addons/website/models/website.py
+++ addons/website/models/website.py
@@ -479,14 +479,6 @@ class Website(models.Model):
         theme_name = kwargs['theme_name']
         theme = request.env['ir.module.module'].search([('name', '=', theme_name)])
         url = theme.button_choose_theme()
-        # button_choose_theme redirects the user with
-        # with_loading and enable_editor but if the
-        # user is on a translated page (e.g /en)
-        # the editor never starts (as there's no edit button)
-        # thus never removes the loading animation
-        # so we redirect the user to the default
-        # language so the editor can initialize properly
-        url['url'] = '/website/lang/default?%s' % urls.url_encode({'r': url['url']})

         # Force to refresh env after install of module
         assert request.env.registry is registry()
@@ -1287,6 +1279,8 @@ class Website(models.Model):
     def button_go_website(self, path='/', mode_edit=False):
         self._force()
         if mode_edit:
+            # force default lang
+            path = url_for(path, self.default_lang_id.code)
             path += '?enable_editor=1'
         return {
             'type': 'ir.actions.act_url',

```

**Comment by detrouxdev** [comment]
I added an extra commit which is an alternative to the fix to make the configurator in the user's language. The issue with the original commit is that it changes behaviours in other part of the website. 
Things like dialog boxes in frontend will be in the user's language instead of being in the frontend language like before. 
The WIP commit only changes with the user_context for the configurator but does it in a "hacky" way.
If anyone following this PR has input on which is the best way to deal with the issue, I'd greatly appreciate it. 🙂

**Comment by JKE-be** [comment]
here my 5 cents
on devrait utiliser la langue default du website plutot que du user...
E.g. https://github.com/odoo/odoo/compare/15.0...odoo-dev:saas-14.4-website-translate-configurator-features-ard-jke

**Comment by detrouxdev** [comment]
Je suis d'accord. Comme c'est un controller frontend ça aurait plus de sens de faire le rendu de la page dans la langue par défaut du nv site web.

Qu'est-ce que t'en penses @rdeodoo ?

**Comment by qsm-odoo** [comment]
@detrouxdev Did you get your answer for the question above? Anyways, not sure what is blocking here but if it is that question: do the simplest solution in stable (is it easier to use the website language?). With YOL's task in master we may want to convert the configurator to a backend action (as I wanted originally btw...) and it will naturally use the user language I guess. Hopefully simplifying code around here.

**Comment by rdeodoo** [comment]
I need to check JKE's branch and move forward with this one.
Thank you for the clear step to reproduces as I [requested](https://github.com/odoo/odoo/pull/80055#issuecomment-1005543274) @detrouxdev :+1: 
I'll try to work on it by Wednesday

**Comment by qsm-odoo** [comment]
Ok nice :+1:

**Comment by detrouxdev** [comment]
> Edit: note that JKE found a bug with OWL leading to issues being fixed here

I think it was fix here d959638e23a92cf880049a760c42da33e5908f41

**Review by rdeodoo — CHANGES_REQUESTED** [review]
A few things here:
1. The first commit had a `[1]` but no markdown link :shrug: 
image

2. The second commit had a `[1]` but a non absolute link..
This is useless, as the markdown link purpose is to make the link clickable outside of github, like in tig/git blame etc in terminal. It needs to be the full link (don't mind it if it goes > 72 char).
Also, written like you did, you would have been better to not use markdown but the first 7-8 (I am not sure) characters of the commit hash directly inside the sentence, as the markdown here doesn't bring value.
But I am sure you wanted to use a full link ;-)
image

3. For the second commit, the step to reproduce is over-complicated. I am not sure if you realized during testing but you can simply have the bug by creating a database with only french language installed. Everything in the configurator will be in french except the features.

4. I don't think I saw anywhere the third bug being mentioned, where the second lang of the website pages were not translated (not inside the configurator, in the website itself).
But it is still there and this case is not solved, see on your runbot: https://watch.screencastify.com/v/j38KtfkfTPlVGFnRDgOV

Note that I pushed the markdown link change myself to quickly r+ but spotted error 4. so.. I'll let you handle that part.

This is getting tricky, too many cases & any change is reverting previous fixes of the PR. Not to mention what could happen with stuff like http pocalypse, owl2 etc.
We really need to introduce tests here, about every flows being fixed, and the one I mentioned which is still broken. Can you have a look? Probably some tours-tests to be sure it works as expected from A to Z, as this is the core feature of the website app.

File: addons/website/tests/test_configurator_translation.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+import odoo.tests
+from odoo.addons.iap.tools.iap_tools import iap_jsonrpc_mocked
+from odoo.tools.misc import mute_logger
+
+@odoo.tests.common.tagged('post_install', '-at_install')
+class TestConfigurator(odoo.tests.HttpCase):
+
+    def _theme_upgrade_upstream(self):
+        # patch to prevent module install/upgrade during tests
```
Comment (rdeodoo) [inline addons/website/tests/test_configurator_translation.py:14]: I was like "Cool, he thought about that", then saw it was a copy pasta.. You can't obviously copy paste a whole file to just add a few lines.. (in this case the if condition on language and the fr case)

You need to move this to the `test_configurator.py` file and use the same `setUp()`.
And what can't be shared inside the setup, you just move it to your test method.

Comment (detrouxdev) [reply addons/website/tests/test_configurator_translation.py:14]: Fair enough. I was under the impression that you wanted the test to be inside website and not under the test_website_modules because it wasn't dependant of all the other website module, but rather just the configurator itself. I will move the file back.

Comment (rdeodoo) [reply addons/website/tests/test_configurator_translation.py:14]: Yes that's true, I didn't notice it was in different module, but then you can just move the setup code from test_website_modules' test to the one you introduce, and makes test_website_modules class inherits from the one you introduce, so we don't end up with twice the same code.

File: addons/website/tests/test_configurator_translation.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+                return []
+            elif '/api/website/2/configurator/custom_resources/' in endpoint:
+                return {'images': {}}
+
+            iap_jsonrpc_mocked()
+
+        with mute_logger('odoo.addons.base.models.ir_translation'):
+            self.env["base.language.install"].create({'lang': 'fr_FR', 'overwrite': True}).lang_install()
+        self.lang_id = self.env.ref('base.lang_fr')
+        self.fr_website_id = self.env['website'].create({'name': "Website Fr", 'default_lang_id': self.lang_id.id})
```
Comment (rdeodoo) [inline addons/website/tests/test_configurator_translation.py:51]:
```suggestion
        self.website_fr = self.env['website'].create({'name': "Website Fr", 'default_lang_id': self.lang_id.id})
```
Variable suffixed by `_id` are supposed to hold.. ID.. Not record.

File: addons/website/tests/test_configurator_translation.py (hunk 8 above)
Comment (detrouxdev) [reply addons/website/tests/test_configurator_translation.py:14]: Any advice on how I could patch the inner funciton `iap_jsonrpc_mocked_configurator` from the original `test_configurator.py`?

Comment (rdeodoo) [reply addons/website/tests/test_configurator_translation.py:14]: I am not too sure to understand what you mean.

The `test_configurator.py` will become a 8 lines file:

```python
# -*- coding: utf-8 -*-
# Part of Odoo. See LICENSE file for full copyright and licensing details.

from unittest.mock import patch

import odoo.tests
from odoo.addons.iap.tools.iap_tools import iap_jsonrpc_mocked
from odoo.addons.website.tests.test_configurator_translation import TestConfigurator


@odoo.tests.common.tagged('post_install', '-at_install')
class TestConfigurator(TestConfigurator):

    def test_01_configurator_flow(self):
        self.start_tour('/web#action=website.action_website_configuration', 'configurator_flow', login="admin")

```

Comment (rdeodoo) [reply addons/website/tests/test_configurator_translation.py:14]:
```diff
diff --git addons/test_website_modules/tests/test_configurator.py addons/test_website_modules/tests/test_configurator.py
index 972b2116d5f..e10cc13df6b 100644
--- addons/test_website_modules/tests/test_configurator.py
+++ addons/test_website_modules/tests/test_configurator.py
@@ -5,42 +5,11 @@ from unittest.mock import patch
 
 import odoo.tests
 from odoo.addons.iap.tools.iap_tools import iap_jsonrpc_mocked
+from odoo.addons.website.tests.test_configurator_translation import TestConfigurator
 
 
 @odoo.tests.common.tagged('post_install', '-at_install')
-class TestConfigurator(odoo.tests.HttpCase):
-
-    def _theme_upgrade_upstream(self):
-        # patch to prevent module install/upgrade during tests
-        pass
-
-    def setUp(self):
-        super().setUp()
-
-        def iap_jsonrpc_mocked_configurator(*args, **kwargs):
-            endpoint = args[0]
-            if endpoint.endswith('/api/website/1/configurator/industries'):
-                return {"industries": [
-                    {"id": 1, "label": "abbey"},
-                    {"id": 2, "label": "aboriginal and torres strait islander organisation"},
-                    {"id": 3, "label": "aboriginal art gallery"},
-                    {"id": 4, "label": "abortion clinic"},
-                    {"id": 5, "label": "abrasives supplier"},
-                    {"id": 6, "label": "abundant life church"}]}
-            elif '/api/website/2/configurator/recommended_themes' in endpoint:
-                return []
-            elif '/api/website/2/configurator/custom_resources/' in endpoint:
-                return {'images': {}}
-
-            iap_jsonrpc_mocked()
-
-        iap_patch = patch('odoo.addons.iap.tools.iap_tools.iap_jsonrpc', iap_jsonrpc_mocked_configurator)
-        iap_patch.start()
-        self.addCleanup(iap_patch.stop)
-
-        patcher = patch('odoo.addons.website.models.ir_module_module.IrModuleModule._theme_upgrade_upstream', wraps=self._theme_upgrade_upstream)
-        patcher.start()
-        self.addCleanup(patcher.stop)
+class TestConfigurator(TestConfigurator):
 
     def test_01_configurator_flow(self):
         self.start_tour('/web#action=website.action_website_configuration', 'configurator_flow', login="admin")
diff --git addons/website/tests/test_configurator_translation.py addons/website/tests/test_configurator_translation.py
index 16cad7d1152..a7cf1bd2e21 100644
--- addons/website/tests/test_configurator_translation.py
+++ addons/website/tests/test_configurator_translation.py
@@ -46,10 +46,6 @@ class TestConfigurator(odoo.tests.HttpCase):
 
             iap_jsonrpc_mocked()
 
-        with mute_logger('odoo.addons.base.models.ir_translation'):
-            self.env["base.language.install"].create({'lang': 'fr_FR', 'overwrite': True}).lang_install()
-        self.lang_id = self.env.ref('base.lang_fr')
-        self.fr_website_id = self.env['website'].create({'name': "Website Fr", 'default_lang_id': self.lang_id.id})
         iap_patch = patch('odoo.addons.iap.tools.iap_tools.iap_jsonrpc', iap_jsonrpc_mocked_configurator)
         iap_patch.start()
         self.addCleanup(iap_patch.stop)
@@ -59,7 +55,14 @@ class TestConfigurator(odoo.tests.HttpCase):
         self.addCleanup(patcher.stop)
 
     def test_01_configurator_translation(self):
+        with mute_logger('odoo.addons.base.models.ir_translation'):
+            self.env["base.language.install"].create({'lang': 'fr_FR', 'overwrite': True}).lang_install()
+        website_fr = self.env['website'].create({
+            'name': "Website Fr",
+            'default_lang_id': self.env.ref('base.lang_fr').id
+        })
+
         # disable configurator todo to ensure this test goes through
         active_todo = self.env['ir.actions.todo'].search([('state', '=', 'open')], limit=1)
         active_todo.update({'state': 'done'})
-        self.start_tour('/website/force/%s?path=%%2Fwebsite%%2Fconfigurator%%3Fdebug%%3Dassets' % self.fr_website_id.id, 'configurator_translation', login='admin')
+        self.start_tour('/website/force/%s?path=%%2Fwebsite%%2Fconfigurator%%3Fdebug%%3Dassets' % website_fr.id, 'configurator_translation', login='admin')
```

Comment (detrouxdev) [reply addons/website/tests/test_configurator_translation.py:14]: I was thinking the other way around (inheriting TestConfigurator in TestConfiguratorTranslation), I miss read, this makes much more sense! Thanks!

File: addons/website/tests/test_configurator_translation.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+        patcher = patch('odoo.addons.website.models.ir_module_module.IrModuleModule._theme_upgrade_upstream', wraps=self._theme_upgrade_upstream)
+        patcher.start()
+        self.addCleanup(patcher.stop)
+
+    def test_01_configurator_translation(self):
+        # disable configurator todo to ensure this test goes through
+        active_todo = self.env['ir.actions.todo'].search([('state', '=', 'open')], limit=1)
+        active_todo.update({'state': 'done'})
+        self.start_tour('/website/force/%s?path=%%2Fwebsite%%2Fconfigurator%%3Fdebug%%3Dassets' % self.fr_website_id.id, 'configurator_translation', login='admin')
```
Comment (rdeodoo) [inline addons/website/tests/test_configurator_translation.py:64]: The test/tour is not crashing without your first commit (about infinite loading).
Do you think you can call a second tour in the same test method to check this case? (which was the only really blocking issue)

File: addons/website/tests/test_configurator_translation.py (hunk 8 above)
Comment (rdeodoo) [reply addons/website/tests/test_configurator_translation.py:14]: :) also maybe chose another class name, not the same one, name it like its file TestConfiguratorTranslation

File: addons/website/tests/test_configurator_translation.py (hunk 10 above)
Comment (detrouxdev) [reply addons/website/tests/test_configurator_translation.py:64]: Ok so I was thinking : the second commit prevent the bug from ever happening. We're always going to have a frontend_lang equal to the website we're going to edit in the second commit because, when we create a website, we go on the configurator page before we can skip it. And going to the configurator sets the correct frontend_lang.

That doesn't mean the first commit is useless however, in fact it should be pack-ported because the issue also happens in 14.0:
 
 - Create a new databse
 - Install website, and add french on website 1
 - Change lang of website 1 to french while visiting
 - Create a new website with english as a language
 - Select a theme
 - The editor won't start even though ?enable_editor=1 is in the address bar

Should I make a PR for 14.0? I'll add tour that reproduce the steps in both PR..

Comment (rdeodoo) [reply addons/website/tests/test_configurator_translation.py:64]: If you can then extract that commit in 14.0 that would be perfect.
It will be added in 15.0 and master during forward port, not needed in this PR then IMO

File: addons/website/tests/test_configurator.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+        patcher = patch('odoo.addons.website.models.ir_module_module.IrModuleModule._theme_upgrade_upstream', wraps=self._theme_upgrade_upstream)
+        patcher.start()
+        self.addCleanup(patcher.stop)
+
+@odoo.tests.common.tagged('post_install', '-at_install')
+class TestConfiguratorTranslation(TestConfiguratorCommon):
+
+    def test_01_configurator_translation(self):
+        with mute_logger('odoo.addons.base.models.ir_translation'):
```
Comment (rdeodoo) [inline addons/website/tests/test_configurator.py:59]: Do we really need this? Why is lang_install raising error?

File: addons/website/static/tests/tours/configurator_translation.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        content: "select a website type",
+        trigger: 'a.o_change_website_type',
+    }, {
+        content: "insert a website industry",
+        trigger: '.o_configurator_industry input',
+        run: 'text ab',
+    }, {
+        content: "select a website industry from the autocomplete",
+        trigger: '.o_configurator_industry ul li a:contains(in fr)',
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/static/tests/tours/configurator_translation.js:26]: Well I guess it works, but I was more speaking about testing one of the sentence like "I want" (less robust if we change text, but would be sure it really tests what we expect).
But you can keep it like that that's already something

**Review by rdeodoo — CHANGES_REQUESTED** [review]
I'll send that to PDE for a quick testing round as this is quite tricky.
Remaining reviews are mostly about test so it won't change anything

File: addons/website/tests/test_configurator.py (hunk 11 above)
Comment (rdeodoo) [reply addons/website/tests/test_configurator.py:59]: Ok I see, I don't like that:
- this might be muting useful error one day if I am not mistaken
- this is just to clean a bit the logs, which are.. logs

At that point, if that's just for cleaning, we might as well hide the remaining log entries about .po, as this is really not relevant in this PR since we create a new language and there is not a single .po file for it obviously. That step is ultra fast since no .po, might as well get rid of that log.

Anyway, that's not a big deal, you can keep it like that, the code where you found this was done by @mart-e so I trust that code ;-)

Comment (mart-e) [reply addons/website/tests/test_configurator.py:59]: I already did things like this (e.g. [`test_export_import`](https://github.com/odoo/odoo/blob/336c9aab3619abb5bfbeb07f87ae014facdefb0b/odoo/addons/test_translation_import/tests/test_term_count.py#L189-L190)). I think it was indeed to avoid too much noise. There is a redundant log in `trans_load` and `_load_module_terms`. Now that the translation system is a bit more mature, we could probably get rid (or lower to debug) one of them.

Comment (rdeodoo) [reply addons/website/tests/test_configurator.py:59]: Thanks for the explanation, that's then indeed to reduce "noise". So @detrouxdev you can keep it like that for this new test in stable :+1:

File: addons/website/static/tests/tours/configurator_translation.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        content: "Loader should be shown",
+        trigger: '.o_theme_install_loader_container',
+        run: function () {}, // it's a check
+    }, {
+        content: "Wait untill the configurator is finished",
+        trigger: 'body.editor_started',
+        timeout: 30000,
+    }, {
+        content: "exit edit mode",
+        trigger: '.o_we_website_top_actions button.btn-primary:contains("Sauver")',
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/configurator_translation.js:57]: See https://github.com/odoo/odoo/pull/83508#discussion_r803809663

File: addons/website/static/tests/tours/configurator_translation.js [hunk 14]
```diff
@@ -0,0 +1,64 @@
+odoo.define('test_website_modules.tour.configurator_translation', function (require) {
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/configurator_translation.js]: `test_website_modules` :thinking: 

You can actually get rid of all that naming stuff if using the new way to declare JS files, As simple as that:

```
/** @odoo-module */

import tour from 'web_tour.tour';

tour.register('configurator_translation', {
```

**Review by rdeodoo — APPROVED** [review]
Thanks, pushed the small test change.
@robodoo r+

**Comment by Julien00859** [comment]
@detrouxdev @rdeodoo @qsm-odoo 

> With the new http-pocalypse coming, I am afraid such behaviors might be lost, as it really impacts those layer.
> Couldn't you write some tests here?
Great call! It indeed did crash in my branch, thank you for the unittest <3

So there is an important difference master vs httpocalypse, in master you update the context's lang according to the frontend lang (request.lang) even in case `frontend_multilang=False`. In my branch the context's lang is only updated in case `frontend_multilang=True`. I note too that I have the same problem with the cookie lang, in my case it is only set in case of `frontend_multilang=True` and in master you always set it.

Which of the two is right? I must confess that I was quite surprised to read lang manipulation in a not-frontend-multilang controller, I'm about sure my assumptions in my branch are wrong and that I should restore the lang in both the context and the cookie, can you confirm it to me?

Also, we are on the verge of merging httpocalypse, this is one problem is the *only* problem left.

**Comment by Julien00859** [comment]
For references:

context lang in master: https://github.com/odoo/odoo/blob/561fee585643cdf3ea3af2ef548b2d5fd1d86a1c/addons/http_routing/models/ir_http.py#L403-L407 (set in case `is_frontend`)
cookie lang in master: https://github.com/odoo/odoo/blob/561fee585643cdf3ea3af2ef548b2d5fd1d86a1c/addons/http_routing/models/ir_http.py#L517-L519 (set in case `is_frontend`)

context/cookie lang in httpocalypse: https://github.com/odoo-dev/odoo/blob/2c7f2e8e7bbb578a9bbf4235a1437c34f6ada9b2/addons/http_routing/models/ir_http.py#L575-L578 (set in case `is_frontend_multilang`)

**Comment by rdeodoo** [comment]
Handled in discord, this will be most likely kept as in current master
